perses/tests/test_hybrid_builder.py

```diff
@@ -99,6 +123,173 @@
+
+    #Now create the compound states with different alchemical states
+    lambda_zero_thermodynamic_state = CompoundThermodynamicState(thermodynamic_state, composable_states=[lambda_zero_alchemical_state])
+    lambda_one_thermodynamic_state = CompoundThermodynamicState(thermodynamic_state, composable_states=[lambda_one_alchemical_state])
+
+    return nonalchemical_zero_thermodynamic_state, nonalchemical_one_thermodynamic_state, lambda_zero_thermodynamic_state, lambda_one_thermodynamic_state
+
+def run_endpoint_perturbation(lambda_thermodynamic_state, nonalchemical_thermodynamic_state, initial_hybrid_sampler_state, mc_move, n_iterations, factory, lambda_index=0):
+    """
+
+    Parameters
+    ----------
+    lambda_thermodynamic_state
+    nonalchemical_thermodynamic_state
+    initial_hybrid_sampler_state
+    mc_move
+    n_iterations
+    """
+    #run an initial minimization:
+    mcmc_sampler = mcmc.MCMCSampler(lambda_thermodynamic_state, initial_hybrid_sampler_state, mc_move)
+    mcmc_sampler.minimize(max_iterations=20)
+    new_sampler_state = mcmc_sampler.sampler_state
+
+    #initialize work array
+    w = np.zeros([n_iterations])
+
+    #run n_iterations of the endpoint perturbation:
+    for iteration in range(n_iterations):
+        mc_move.apply(lambda_thermodynamic_state, new_sampler_state)
+
+        #compute the reduced potential at the new state
+        hybrid_context, integrator = cache.global_context_cache.get_context(lambda_thermodynamic_state)
+        new_sampler_state.apply_to_context(hybrid_context, ignore_velocities=True)
+        hybrid_reduced_potential = lambda_thermodynamic_state.reduced_potential(hybrid_context)
+
+        #generate a sampler state for the nonalchemical system
+        if lambda_index == 0:
+            nonalchemical_positions = factory.old_positions(new_sampler_state.positions)
+        elif lambda_index == 1:
+            nonalchemical_positions = factory.new_positions(new_sampler_state.positions)
+        else:
+            raise ValueError("The lambda index needs to be either one or zero for this to be meaningful")
+
+        nonalchemical_sampler_state = SamplerState(nonalchemical_positions, box_vectors=new_sampler_state.box_vectors)
+
+        #compute the reduced potential at the nonalchemical system as well:
+        nonalchemical_context, integrator = cache.global_context_cache.get_context(nonalchemical_thermodynamic_state)
+        nonalchemical_sampler_state.apply_to_context(nonalchemical_context, ignore_velocities=True)
+        nonalchemical_reduced_potential = nonalchemical_thermodynamic_state.reduced_potential(nonalchemical_context)
+
+        w[iteration] = nonalchemical_reduced_potential - hybrid_reduced_potential
+
+    [t0, g, Neff_max] = timeseries.detectEquilibration(w)
+    w_burned_in = w[t0:]
+
+    [df, ddf] = pymbar.EXP(w_burned_in)
+    ddf_corrected = ddf * np.sqrt(g)
+
+    assert ddf_corrected < 3.0
```

Comment:
Perhaps the threshold should be a constant specified as an optional argument for whatever function does the error checking?

Reply:
Perhaps this method should just return `ddf_corrected` and the calling function should test the uncertainty is below a specified threshold, returning helpful error statements if it isn't?